compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -3500,25 +3500,29 @@ private void validateAnnotationAttachmentExpr(BLangAnnotationAttachment annAttac
             return;
         }
 
-
         // At this point the type is a subtype of  map<anydata>|record{ anydata...; } or
-        // map<anydata>[]|record{ anydata...; }[], thus an expression is required.
-        if (annAttachmentNode.expr == null) {
+        // map<anydata>[]|record{ anydata...; }[]. For map<anydata>[]|record{ anydata...; }[] an expression is
+        // required while map<anydata>|record{ anydata...; } can be empty.
+        if (annAttachmentNode.expr == null && annotationSymbol.attachedType != null &&
+                !(annotationSymbol.attachedType.type.tag == TypeTags.MAP ||
+                        annotationSymbol.attachedType.type.tag == TypeTags.RECORD)) {
             this.dlog.error(annAttachmentNode.pos, DiagnosticErrorCode.ANNOTATION_ATTACHMENT_REQUIRES_A_VALUE,
-                            annotationSymbol);
+                    annotationSymbol);
             return;
         }
 
-        BType annotType = annotationSymbol.attachedType.type;
-        this.typeChecker.checkExpr(annAttachmentNode.expr, env,
-                                   annotType.tag == TypeTags.ARRAY ? ((BArrayType) annotType).eType : annotType);
+        if (annAttachmentNode.expr != null) {
```

Comment:
If `annAttachmentNode.expr` is null don't we have to do some semantic validation to check if the empty mapping constructor expression is valid for that particular type?

For example, IMO, the following should result in a compilation error, since the record type has a required field. Right now, this seems to be allowed at compile-time resulting in invalid values at runtime.

This crashes at runtime.
```ballerina
type Record record {| int[] i; |};

annotation Record x on function;

@x
function func() {
} 

public function main() { 
   Record rec = <Record> (typeof func).@x; 
   int i = rec.i[0];
}
```

```log
[2021-03-03 10:15:03,754] SEVERE {b7a.log.crash} - null 
java.lang.NullPointerException 
    at test.main(test.bal:22) 
    at $_init.$lambda$main$(.) 
    at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:571) 
    at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:301) 
    at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:269) 
    at java.base/java.lang.Thread.run(Thread.java:834)
```